### PR TITLE
PubSub v0.11.0

### LINKED ISSUE #55: PubSub
Hello!

I tried to send and receive messages locally but sending "hello world" I received "aGVsbG8gd29ybGQ=".
If I use the CLI ipfs pubsub sub <topic> It work as expected which makes me thing there's a bug somewhere.

I'll look around and see if I can fix it.

Otherwise great API!

### PR DESCRIPTION
**Pull request by SionoiS**
Hello!

This PR update the pubsub to go-ipfs v0.11.0

Makes #67 obsolete & Close #55 & #92

The Pubsub example works for 1 message then panic. IDK why... something to do with json deserialization in the request. Would need to be fixed before merging.

### CODE REVIEW

**Comment by SionoiS** [comment]
I changed pubsub messages types. Deserialization now result in proper types for the various fields.

**Comment by ferristseng** [comment]
Thanks! Will try to find some time to test this myself.

> I changed pubsub message types. Deserialization now result in proper types for the various fields.

Is that the reason for the failing test?

**Comment by SionoiS** [comment]
Can't be since it does work once then panic with the same message. Also, the panic is when sending the message, not receiving, which is even weirder.

I tried to find the problem but I'm stumped...

**Comment by ferristseng** [comment]
I made a PR to resolve the issue! Let me know if that fixes it

File: ipfs-api-prelude/src/api.rs [hunk 1]
```diff
@@ -2024,15 +2024,27 @@ pub trait IpfsApi: Backend {
...
+        topic: T,
+        data: R,
+    ) -> Result<response::PubsubPubResponse, Self::Error>
+    where
+        T: AsRef<[u8]>,
+        R: 'static + Read + Send + Sync,
+    {
+        use multibase::{encode, Base};
+
+        let topic = encode(Base::Base64Url, topic);
```
Comment (ferristseng) [inline ipfs-api-prelude/src/api.rs:2041]: Would it be possible to do this in serde, similar to how you're handling deserialization?

File: ipfs-api-prelude/src/response/pubsub.rs [hunk 2]
```diff
@@ -27,15 +30,60 @@ pub type PubsubPubResponse = ();
...
+    D: Deserializer<'de>,
+{
+    let data: &str = Deserialize::deserialize(deserializer)?;
+
+    let (_, data) = decode(data).expect("Multibase Decoding");
+
+    Ok(data)
+}
+
+fn deserialize_seqno_field<'de, D>(deserializer: D) -> Result<u64, D::Error>
```
Comment (ferristseng) [inline ipfs-api-prelude/src/response/pubsub.rs:56]: There is a `serde` file where I have a few other deserializae methods. Could you move these there?

File: ipfs-api-prelude/src/response/pubsub.rs [hunk 3]
```diff
@@ -27,15 +30,60 @@ pub type PubsubPubResponse = ();
...
+where
+    D: Deserializer<'de>,
+{
+    let seqno: &str = Deserialize::deserialize(deserializer)?;
+
+    let (_, seqno) = decode(seqno).expect("Multibase Decoding");
+
+    let seqno = seqno.try_into().expect("64 bits Type");
```
Comment (ferristseng) [inline ipfs-api-prelude/src/response/pubsub.rs:64]: Can these methods return an Err instead of panicking?

File: ipfs-api-prelude/src/api.rs (hunk 1 above)
Comment (SionoiS) [reply ipfs-api-prelude/src/api.rs:2041]: Do I add a serde.rs in requests for the serializer impl ?

File: ipfs-api-prelude/src/response/pubsub.rs (hunk 3 above)
Comment (SionoiS) [reply ipfs-api-prelude/src/response/pubsub.rs:64]: I could not remove all possible panic.

**Comment by SionoiS** [comment]
Let me know what you think!

**Comment by ferristseng** [comment]
Looks awesome! Thank you!

**Pull request merged**